### PR TITLE
Add rate and channels to HDA config

### PR DESCRIPTION
**Pull request by bardliao**
FW needs to know rate and channels for HDA dai.

This PR should not be merged before #1943

### CODE REVIEW

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -3064,7 +3070,7 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
 	config->hdr.size = size;
 
 	/* get any bespoke DAI tokens */
-	ret = sof_parse_tokens(scomp, config, hda_tokens,
+	ret = sof_parse_tokens(scomp, &config->hda, hda_tokens,
```
Comment (plbossart) [inline sound/soc/sof/topology.c:3110]: I wonder if this belongs in a separate patch. This was obviously never right, so maybe recognize it as a fix, then add new things?

Comment (bardliao) [reply sound/soc/sof/topology.c:3110]: > I wonder if this belongs in a separate patch. This was obviously never right, so maybe recognize it as a fix, then add new things?

Done

**Comment by slawblauciak** [comment]
I wonder if we should add the FIFO size, since that is configurable by the kernel. Right now it's probably not needed for anything, but we're already extending the ABI and the FW does propagate the FIFO size information.

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -3064,7 +3064,7 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
 	config->hdr.size = size;
```
Comment (kv2019i) [inline sound/soc/sof/topology.c:3107]: @bardliao Can you add a mention in commit message that as "hda_tokens" array was empty, this error has been harmless.

File: include/sound/sof/dai-intel.h [hunk 3]
```diff
@@ -88,6 +88,8 @@ struct sof_ipc_dai_ssp_params {
 struct sof_ipc_dai_hda_params {
 	struct sof_ipc_hdr hdr;
 	uint32_t link_dma_ch;
+	uint32_t rate;
```
Comment (kv2019i) [inline include/sound/sof/dai-intel.h:91]: @bardliao Please add a short summary of why the change is done why it doesn't break the ABI. This doesn't have to be detailed explanation, but I think a bit more is warranted in the kernel side commit. Maybe something like: "FW interface for HDA DAI parameters was extended with information on sampling rate and channel count in version 3.16. Align kernel header with the FW change. This change is backwards compatible. Old firmware will ignore the values."

File: include/uapi/sound/sof/tokens.h [hunk 4]
```diff
@@ -126,4 +126,8 @@
 #define SOF_TKN_MUTE_LED_USE			1300
 #define SOF_TKN_MUTE_LED_DIRECTION		1301
 
+/* HDA */
```
Comment (kv2019i) [inline include/uapi/sound/sof/tokens.h:133]: Typo in summary line: rate amd channels -> rate and channels

Comment (kv2019i) [reply include/uapi/sound/sof/tokens.h:133]: I wonder whether it makes sense to keep 2nd and 3rd patches separate. The first one doesn't really make much sense on its own. And I don't think it's worth having the same explanation again in this commit, on why this is done. 

If you keep these separate, maybe something more useful to commit description than 
"FW needs to know rate and channels." 
Maybe just add "Added support for rate and channel tokens for HDA DAI. Added in topology interface 3.16."

**Comment by bardliao** [comment]
Thanks @kv2019i , PR updated
